htlcswitch/switch.go

```diff
@@ -353,35 +340,80 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
 	return nil
 }
 
+// GetPaymentResult returns the the result of the payment attempt with the
+// given paymentID. The method returns a channel where the payment result will
+// be sent when available, or an error is encountered. If the paymentID is
+// unknown, ErrPaymentIDNotFound will be returned.
+func (s *Switch) GetPaymentResult(paymentID uint64,
+	deobfuscator ErrorDecrypter) (<-chan *PaymentResult, error) {
+
+	s.pendingMutex.Lock()
+	payment, ok := s.pendingPayments[paymentID]
+	s.pendingMutex.Unlock()
+
+	if !ok {
+		return nil, ErrPaymentIDNotFound
+	}
+
+	resultChan := make(chan *PaymentResult, 1)
+
+	// Since the payment was known, we can start a goroutine that can
+	// extract the result when it is available, and pass it on to the
+	// caller.
+	s.wg.Add(1)
+	go func() {
```

Comment:
`GetPaymentID` needs to return for the router to know whether the payment ID existed (also to not deadlock the interaction router<->switch on shutdown).

Reply:
But `GetPaymentResult` without the goroutine would still return early with `ErrPaymentIDNotFound` if it doesn't exist?

Does a dead lock happen if both router and switch wait for closure of the same quit chan?

Reply:
making the method return a result async makes it more natural that this is a long-running process. The deadlock happens if the router is waiting for the method to return on shutdown, while the switch will be shutdown after the router.

Reply:
It was sort of pre-existing, but it wouldn't deadlock on shutdown. Since `router.SendPayment` wouldn't return when shutting down, but the goroutines calling it aren't waitgrouped.

Passing in a quit channel is possible, but not the nicest API imo, as we are not actually cancelling the HTLC or anything, only triggering the call to return.

Having the decryption being done in the router was what the previous iteration of the PR did, but the diff became much larger (especially because the tests and mocks must be changed to return an encrypted error), and we have to pass the error metadata to the router, so we decided to go this route. See: https://github.com/lightningnetwork/lnd/commit/d9ccc3431f47c120d71049ea04f462e4f6d9b7e2

In short terms, it is mostly done to keep the boundary close to the current one, where the `ForwardingError` is returned directly to the router. I do agree it is not optional, but it can be changed later.

Reply:
Ok, my suggestion was not to move decryption into the router, but to just to make it a separate call from router to switch. So `switch.SendHTLC` -> `switch.GetPaymentResult` -> `switch.Decrypt` (so slightly different from previous iteration which I remember). But can understand that in this variation the diff in tests may still be large.